spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/TaskMetrics.java

```diff
@@ -125,11 +125,11 @@ public void onTaskFailed(Throwable throwable) {
 
 	public void onTaskEnd(TaskExecution taskExecution) {
 		if (this.scope != null) {
-			if (this.scope != null) {
-				this.taskObservationContext.getTaskExecution().setExitCode(taskExecution.getExitCode());
-				this.scope.getCurrentObservation().stop();
-				this.scope = null;
-			}
+			this.taskObservationContext.getTaskExecution().setExitCode(taskExecution.getExitCode());
+			this.scope.getCurrentObservation().stop();
+			this.scope = null;
+			this.registry.getCurrentObservation().stop();
+			this.registry.setCurrentObservation(null);
```

Comment:
Why are you nulling this in this line? Line 129 should be doing this with the `this.scope.close()` method. In general when `onTaskStart` creates an observation then `onTaskEnd` should stop it and remove it from thread local via `this.scope.close()`